### PR TITLE
Fix: parse to non-default node storage and borrow optimization from serd

### PR DESCRIPTION
**Pull request by liss-h**

### CODE REVIEW

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp [hunk 1]
```diff
@@ -15,7 +21,7 @@ namespace rdf4cpp::rdf::parser {
 static size_t istream_read(void *buf, [[maybe_unused]] size_t elem_size, size_t count, void *voided_self) noexcept {
     assert(elem_size == 1);
 
-    auto *self = reinterpret_cast<std::istream *>(voided_self);
+    auto *self = static_cast<std::istream *>(voided_self);
```
Comment (bigerl) [inline src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp:24]: Just read-up upon this. Seems much better to use static_cast everywhere to cast from voided to concrete pointers.

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp [hunk 2]
```diff
@@ -47,17 +53,15 @@ IStreamQuadIterator::IStreamQuadIterator(void *stream,
                                          ReadFunc read,
                                          ErrorFunc error,
                                          flags_type flags,
-                                         state_type *state,
-                                         storage::node::NodeStorage node_storage) noexcept
-    : impl{std::make_unique<Impl>(stream, read, error, flags, state, std::move(node_storage))} {
+                                         state_type *state) noexcept
```
Comment (bigerl) [inline src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp:56]: I'll use this spot to phrase the general problem: the configuration of the node_storage was moved to the state at some point. But the constructor parameter remained and was still set. So, even when it was set nodes were written to the default node storage. 

Correct?

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp [hunk 3]
```diff
@@ -97,4 +101,19 @@ bool IStreamQuadIterator::operator!=(std::default_sentinel_t) const noexcept {
...
 
+FILE *fopen_fastseq(char const *path, char const *mode) noexcept {
+    // inspired by <serd/system.c> (serd_fopen)
+
+    FILE *fd = fopen(path, mode);
+    if (fd == nullptr) {
+        return fd;
+    }
+
+#if _POSIX_C_SOURCE >= 200112L
```
Comment (bigerl) [inline src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp]: the header guard `#if __has_include(<fcntl.h>)` and this should have the same condition. Safest would be to use the conjunction.

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp [hunk 4]
```diff
@@ -97,4 +101,19 @@ bool IStreamQuadIterator::operator!=(std::default_sentinel_t) const noexcept {
...
+    (void) posix_fadvise(fileno(fd), 0, 0, POSIX_FADV_SEQUENTIAL | POSIX_FADV_NOREUSE);
```
Comment (bigerl) [inline src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp]: We could even think about adding `POSIX_FADV_WILLNEED` (non-blocking)

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp [hunk 5]
```diff
@@ -97,4 +101,19 @@ bool IStreamQuadIterator::operator!=(std::default_sentinel_t) const noexcept {
     return !this->is_at_end();
 }
 
+FILE *fopen_fastseq(char const *path, char const *mode) noexcept {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp:104]: Did you measure the difference? (non-blocking)

**Review by bigerl — CHANGES_REQUESTED** [review]
thx, looks good.  Just some clarifications in minor things.

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp (hunk 2 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp:56]: Yes

File: src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp (hunk 5 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp:104]: No I was just looking up how serd opens files

Comment (liss-h) [reply src/rdf4cpp/rdf/parser/IStreamQuadIterator.cpp:104]: Seemed like a potentially useful addition

**Review by bigerl — APPROVED** [review]

**Pull request merged**